### ISSUE TITLE
Enhancement: Group rules by sass extension, be able to pass in hashes of rules optionally only possible for either syntax

### ISSUE THREAD

**Issue by lozandier** [sasstools/sass-lint#66]
If possible, there seems to be value of explicit options or instructions of having different rules—or a different file  or hash of rules—for `sass` files vs.  `scss` files.

This also mirrors the capabilities of linters such as ESLint. The simplest thing that seemingly should be allowed is two tasks can be ran separately for either syntax

In Gulp 3, it would sorta look like the following: 

````js 
const gulp = require( `gulp` );
const sassLint = require( `gulp-sass-lint` );

const sassLiintTaskDetails = ()=> {

  return gulp.src( `**/*.sass` )
    .pipe( sassLint({
       // A rulesPath property can be added that a rules property overrites
       rules: {
         indentation: [ 2, { size: 2 } ], 
       }
    }) )
    .pipe( sassLint.format() )
    .pipe( sassLint.failOnError() );

};

const scssLiintTaskDetails = ()=> {

  return gulp.src( `**/*.scss` )
    //  Default continues to look for `.sass-lint.yml` (in addition to ideally a json formatted .sass-lint file?) if no hash is passed in
    .pipe( sassLint() )
    .pipe( sassLint.format() )
    .pipe( sassLint.failOnError() );

};

gulp.task( `sass-lint`, sassLintTaskDetails );
gulp.task( `scss-lint`, scssLintTaskDetails );
````

In the meantime, while in theory rules that *can't* be applied to files with the `sass` extension (& the opposite for files with the `scss` will be ignored within a single `.sass-lint.yml` file being used within a codebase using both, that doesn't seem to be the case for me so far (Ran the task on environment using El Capitan, `Node v0.12.7;` I have iojs 1.x, 2.x, 3.x in my system I can quickly report back having different outcomes):
<img width="1358" alt="screen shot 2015-09-01 at 2 03 11 am" src="https://cloud.githubusercontent.com/assets/1223224/9597313/68c1ef56-504e-11e5-9be3-4226f1365202.png">

As you can see, a random rule in random dir I know incompatible with `sass` but not `scss` causes problems. 

Furthermore, it seems to impact the expected behavior of other rules, though I'll leave the details of the `sass-lint.yml` here as well to see if it's a subtle syntax error with my rule file:
<img width="1440" alt="screen shot 2015-09-01 at 2 32 07 am" src="https://cloud.githubusercontent.com/assets/1223224/9597646/e37e8328-5051-11e5-8685-8103b388c1c7.png">

## Environment Details
Node: `v0.12.7` (todo is to run tasks on a project with iojs `v3.0.0` that *doesn't* have other deps not allow that to happen)
OS: El Capitan (OS X 10.11)
`sass-lint` version: `1.1`
`gulp-sass-lint` version: `1.0.1` 

## `.sass-lint.yml` used
```yml
files:
  include: '**/*.s+(a|c)ss'
  ignore:
rules:
  # Extends
  extends-before-mixins: 1
  extends-before-declarations: 1
  placeholder-in-extend: 1

  # Mixins
  mixins-before-declarations: 1

  # Line Spacing
  one-declaration-per-line: 1
  empty-line-between-blocks: 1
  single-line-per-selector: 1

  # Disallows
  no-debug: 1
  no-empty-rulesets: 1
  no-extends: 1
  no-ids: 1
  no-important: 1
  no-warn: 1

  # Style Guide
  leading-zero: 1
  nesting-depth: 2
  property-sort-order: 0

  indentation:
    - 1
    -
    - size: 2

  quotes:
    - 1
    - style: double

  # Inner Spacing
  space-after-comma: 1
  space-before-colon: 1
  space-after-colon: 1
  space-before-brace: 1
  space-before-bang: 1
  space-after-bang: 1
  space-between-parens: 1

  # Final Items
  final-newline: 0
  # Until all files are SCSS, this needsto be zero to not cause errors
  trailing-semicolon: 0

```

**Comment by lozandier** [comment]
When it comes to the overriding part, I'd say `Object.assign()` or the equivalent w/ `lodash` or some other utilities library w/ a *mixin* func would work when the `yml` file is transformed to a Object literal that could be merged in with user-specified rules or files in JSON whenever necessary.

**Comment by Snugug** [comment]
If you are having rules return false-positives for the `sass` syntax or incorrectly try and parse the `sass` syntax, please file individual bugs for that. The rules should be smart enough to adapt to that syntax, and do in most cases (for instance, trailing semicolon ignores the `sass` syntax entirely).

If you'd like separate rules for `sass` and `scss`, you can pass custom configs to each task you have. 

And yes, unknown rules currently do error out.

**Comment by lozandier** [comment]
@Snugug Regarding the latter, are there examples of that, it's currently unclear that functionality. 

When it comes to unknown rules, what aspects of that gotcha is contributing to errors that propagate with the attempt of running the tasks I used file this issue.

I didn't think it was a matter of false positives, but unexpected gotchas with the arrangement of rules.

**Comment by lozandier** [comment]
@Snugug In particular, would the custom configs allow you to define `rulePaths` & explicit rules override to the task itself not unlike what I depicted above? 

The version of `gulp-sass-lint` publically viewable on Github doesn't currently explicitly document a custom config being possible & what can be passed into it.

**Comment by Snugug** [comment]
Issues with the gulp plugin should be directed to that repo

**Comment by lozandier** [comment]
Okay, understood. Thought it was a wrapper to the existing flags of
`sass-lint`.

On Tue, Sep 1, 2015 at 8:16 AM, Sam Richard <notifications@github.com>
wrote:

> Issues with the gulp plugin should be directed to that repo
>
> > On Sep 1, 2015, at 8:12 AM, Kevin Lozandier <notifications@github.com>
> wrote:
> >
> > @Snugug In particular, would the custom configs allow you to define
> rulePaths & explicit rules override to the task itself not unlike what I
> depicted above?
> >
> > The version of gulp-sass-lint publically viewable on Github doesn't
> currently explicitly document a custom config being possible & what can be
> passed into it.
>
> >
> > —
> > Reply to this email directly or view it on GitHub.
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/sasstools/sass-lint/issues/66#issuecomment-136692609>.
>



-- 
Kevin Lozandier
lozandier@gmail.com <lozandier@gmail.com>
